mountpoint-s3/src/fs.rs

```diff
@@ -991,7 +1073,61 @@ where
         //   process. In many cases, the child will then immediately close (flush) the duplicated
         //   file descriptors. We will not complete the upload if we can detect that the process
         //   invoking flush is different from the one that originally opened the file.
-        self.complete_upload(fh, true, Some(pid)).await
+        //
+        // The same for read path. We want to stop the prefetcher and decrease the reader count
+        // as soon as users close a file descriptor so that we don't block users from doing other
+        // operation like overwrite the file.
+        let file_handle = {
+            let file_handles = self.file_handles.read().await;
+            match file_handles.get(&fh) {
+                Some(handle) => handle.clone(),
+                None => return Err(err!(libc::EBADF, "invalid file handle")),
+            }
+        };
+        logging::record_name(file_handle.inode.name());
+        let mut state = file_handle.state.lock().await;
+        match &mut *state {
+            FileHandleState::Created { lookup, flags, pid } => {
+                // This happens when users close a file without calling any read() or write(),
+                // since we don't know what type of handle it would be we need to consider what
+                // to do next for both cases.
+                // * if the file is new or opened in truncate mode, we know it must be a write
+                //   handle so we can start an upload from here.
```

Comment:
Shouldn't we also complete the upload, like we do in `fsync`?

Reply:
Actually no, because the file is still empty when when start the upload and we want to ignore empty files on flush from that special case with multiple file descriptors.

Reply:
I see. Do we even need to start the upload in that case? Is there any difference with just letting `release` do it?

Reply:
Just worried that we have similar (but subtly different) logic in a few places.

Reply:
The upload will be started a bit earlier but there should be no other difference. I can remove the logic here and let `release` handle it.